Dev

uriwin opened the pull request:
HI,
hope this is good

uri-win committed:
added prject code

uri-win committed:
Print equipment details and summary for each team member

noam-toledano commented on src/main/java/Team/TeamMember.java:
```diff
@@ -0,0 +1,61 @@
+        TechnologyEquipmentType.add(equipment);
+    }
+
+    public void printEquipment() {
```
SRP- `teamMember` should not be responsible of printing. Instead, it should return something that will be handled elsewhere

noam-toledano commented on src/main/java/Team/TeamMember.java:
```diff
@@ -0,0 +1,61 @@
+        String equipmentsDetails = "";
+        String equipmentSummary = "";
+
+        System.out.println(this.name + " from team " + this.teamName + " - equipments details:");
```
Coupling- this makes the `teamMember` coupled to console output. what if you wanted to switch to a GUI later down the line?

noam-toledano commented on src/main/java/Team/TeamMember.java:
```diff
@@ -0,0 +1,61 @@
+        this.name = name;
+        this.teamName = teamName;
+        memberUniformEquipment = new HashMap<String, List<UniformEquipment>>();
+        memberTechnologyEquipment = new HashMap<String, List<TechnologyEquipment>>();
```
OCP- adding a third kind of equipment requires you to go through the whole code and changing many classes. also- what will you do with equipment that is both a piece of clothing and technology (such as a smartwatch)?

noam-toledano commented on src/main/java/Team/TeamLeader.java:
```diff
@@ -0,0 +1,23 @@
+        this.teamMembers.add(this);
+    }
+
+    public void addTeamMember(TeamMember newTeamMember){
```
DIP- classes should be based on abstractions (such as interfaces) and not on specific implementations

noam-toledano commented on src/main/java/Team/TeamMember.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    public void addUniformEquipment(UniformEquipment equipment) {
+        List<UniformEquipment> newUniformEquipmentType = memberUniformEquipment.get(equipment.equipmentType());
```
is `newUniformEquipmentType` an equipment type or a `List`?  this is a confusing name. also, if the equipment's type already exists this isn't new, so the word `new` should not necessarily be used

noam-toledano commented on src/main/java/Team/TeamMember.java:
```diff
@@ -0,0 +1,61 @@
+
+    public void printEquipment() {
+        String equipmentsDetails = "";
+        String equipmentSummary = "";
```
for building a string out of multiple substrings it's better to use `StringBuilder`

noam-toledano commented on src/main/java/equipments/Equipment.java:
```diff
@@ -0,0 +1,6 @@
+package equipments;
+
+public interface Equipment {
+    public String equipmentType();
```
methods should contain actions in their name- change this to `getEquipmentType`

uriwin replied:
done.

noam-toledano commented on src/main/java/equipments/EquipmentStatus.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.List;
+
+public class EquipmentStatus {
+    public static <T extends Equipment> String teamMemberEquipmentDetails(HashMap<String, List<T>> memberEquipment) {
```
avoid using `static`- it makes it harder to change implementations later down the line and is impossible to unit-test. there is nothing wrong with creating an instance of a functional class (it's better for the garbage collector too)

noam-toledano replied:
also, why use `T extends Equipment` instead of just `Equipment`?

noam-toledano commented on src/main/java/equipments/technology/InternetComputer.java:
```diff
@@ -0,0 +1,12 @@
+        this.technologyEquipmentType = "Internet Computer";
+    }
+    public String equipmentType() {
+        return "Internet Computer";
```
Code dulpication- why is this defined both in `this.technologyEquipmentType` and in `equipmentType()`?

noam-toledano commented on src/main/java/equipments/technology/TechnologyEquipmentFactory.java:
```diff
@@ -0,0 +1,13 @@
+
+public class TechnologyEquipmentFactory {
+    public TechnologyEquipment getTechnologyEquipment(String technologyEquipmentType, String IP){
+        if(technologyEquipmentType.equalsIgnoreCase("Voip"))
```
for values of a predefined list, it's better to use `enum`s instead of `string`s. a string is very fragile (a small typo could cause a bug)

noam-toledano commented on src/main/java/equipments/technology/TechnologyEquipmentFactory.java:
```diff
@@ -0,0 +1,13 @@
+    public TechnologyEquipment getTechnologyEquipment(String technologyEquipmentType, String IP){
+        if(technologyEquipmentType.equalsIgnoreCase("Voip"))
+            return new Voip(IP);
+        else if (technologyEquipmentType.equalsIgnoreCase("Internet computer"))
```
consider using a `switch case` instead of these chained `if`s

noam-toledano commented on src/main/java/equipments/uniform/Harmonit.java:
```diff
@@ -0,0 +1,16 @@
+
+import equipments.Equipment;
+
+public class Harmonit extends UniformEquipment {
```
*Hermonit (like mount Hermon)

noam-toledano commented on src/main/java/equipments/uniform/Harmonit.java:
```diff
@@ -0,0 +1,16 @@
+
+    @Override
+    public String equipmentType() {
+        return this.uniformEquipmentType;
```
is this logic unique to a Hermonit, or could it be in a parent class?

uri-win committed:
changed to switch instead of if in TechnologyEquipmentFactory.java UniformEquipmentFactory.java

equipmentType() now is a function in the abstract classes.
changed name from Harmonit to Hermonit.

uri-win committed:
added enum for UniformEquipmentTypes.java TechnologyEquipmentTypes.java.

use enum instead of string when creating objects in UniformEquipmentFactory.java TechnologyEquipmentFactory.java

uri-win committed:
changed equipmentType() to getEquipmentType().

instead of one function that print the TeamMember Equipment details and summary, created two functions:
getMemberEquipmentDetails, getMemberEquipmentSummary.

uri-win committed:
functions in EquipmentStatus.java are not static anymore.

Removed generic <T> (no need), and now the functions get input of HashMap<String, List<Equipment>> instead.
TeamMember have memberEquipments instead of memberUniformEquipment and memberTechnologyEquipment for OCP.

uri-win committed:
added ITeamMember and AbstractTeamMember for DIP.

rename TechnologyEquipment.java to AbstractTechnologyEquipment because of conventions (same goes for AbstractUniformEquipment and IEquipment).

uri-win committed:
changed vars names for conventions.
